pages/api/predict/roi.ts

```diff
@@ -0,0 +1,243 @@
+      const query = queryBuilderFn(chunkStart, chunkEnd);
+      const response = await mechGraphClient.request(query);
+      const data = extractor(response);
+
+      // If items are less than limit, there are no more items in this range
+      if (data.length < LIMIT) {
+        return data;
+      }
+
+      // If there are more items, we split the duration into two halves to fetch them in parallel manner
+      const mid = Math.floor((chunkStart + chunkEnd) / 2);
+
+      // Prevent infinite loops if multiple items share the exact same second
+      if (mid <= chunkStart) {
+        return data;
+      }
+
+      const [left, right] = await Promise.all([
+        fetchRecursive(chunkStart, mid),
+        fetchRecursive(mid, chunkEnd),
+      ]);
+
+      return [...left, ...right];
+    } catch (err) {
+      console.error(`Error fetching range ${chunkStart}-${chunkEnd}`, err);
+      throw err;
+    }
+  };
+
+  const initialPromises = [];
+  for (let i = 0; i < chunkCount; i++) {
+    const chunkStart = start + i * initialChunkSize;
+    const chunkEnd = Math.min(end, chunkStart + initialChunkSize);
+
+    initialPromises.push(fetchRecursive(chunkStart, chunkEnd));
+  }
+
+  const results = await Promise.all(initialPromises);
+  const flatData = results.flat();
+
+  // Deduplicate
+  const uniqueMap = new Map();
+  flatData.forEach((item) => {
+    if (item.id) uniqueMap.set(item.id, item);
+  });
+
+  return Array.from(uniqueMap.values());
+}
+
+const fetchMechRequests = async (marketOpenTimestamp: number) => {
+  const today = getMidnightUtcTimestampDaysAgo(0);
+  const lastFourDaysRequests = await recursivelyFetchQueryData({
+    queryBuilderFn: getMechRequestsRangeQuery,
+    extractor: (response: MechRequestsResponse) => response.requests,
+    start: marketOpenTimestamp,
+    end: today,
+  });
+
+  return lastFourDaysRequests;
+};
+
+const fetchRoi = async () => {
+  try {
+    const marketOpenTimestamp = getMidnightUtcTimestampDaysAgo(
+      PREDICT_MARKET_DURATION_DAYS,
+    );
+
+    // Request fees, payouts, staking rewards and olas price
+    const [
+      totalRequestsResult,
+      marketsAndBetsResult,
+      totalRewardsResult,
+      olasInUsdPriceResult,
+      lastFourDaysRequests,
+    ] = await Promise.all([
+      mechGraphClient.request(totalMechRequestsQuery),
+      predictAgentsGraphClient.request(
+        getMarketsAndBetsQuery(marketOpenTimestamp),
+      ),
+      STAKING_GRAPH_CLIENTS.gnosis.request(stakingGlobalsQuery),
+      fetch(COINGECKO_OLAS_IN_USD_PRICE_URL),
+      fetchMechRequests(marketOpenTimestamp),
+    ]);
+
+    const olasInUsdPrice = await olasInUsdPriceResult.json();
+    const olasInUsdPriceInEth = BigInt(
+      Math.floor(Number(olasInUsdPrice[OLAS_ADDRESS]?.usd || 0) * 1e18),
+    );
+
+    const totalMechRequests = (totalRequestsResult as TotalMechRequestsResponse)
+      .global.totalRequests;
+    const openMarketTitles = (
+      marketsAndBetsResult as MarketsAndBetsResponse
```

Comment:
can we maybe put all types to Promise.all like 
```
await Promise.all<
      [
        TotalMechRequestsResponse,
        MarketsAndBetsResponse,
        StakingGlobalsResponse,
        ...
      ]
    >
    ```
not to repeat aliases? can do "any" somewhere with todos (to fix later when other files are migrated)